aten/src/ATen/native/ufunc/sub.h

```diff
@@ -0,0 +1,30 @@
+// TODO: this is a pessimization
+
+#pragma once
+
+#include <c10/macros/Macros.h>
+
+#if !defined(__CUDACC__) && !defined(__HIPCC__)
+#include <ATen/cpu/vec/functional.h>
+#include <ATen/cpu/vec/vec.h>
+#endif
+
+namespace at {
+namespace native {
+namespace ufunc {
+
+template <typename T>
+C10_HOST_DEVICE T sub(T self, T other, T alpha) __ubsan_ignore_undefined__ {
+  return self - alpha * other;
+}
+
+#if !defined(__CUDACC__) && !defined(__HIPCC__)
+using vec::Vectorized;
+template <typename T>
+C10_HOST_DEVICE Vectorized<T> sub(Vectorized<T> self, Vectorized<T> other, Vectorized<T> alpha) __ubsan_ignore_undefined__ {
+  // TODO: this is a pessimization
+  return vec::fmadd(other, alpha.neg(), self);
```

Comment:
what makes this a pessimization?

Reply:
In the original code, the alpha scalar is negated before we enter the inner loop, see https://github.com/pytorch/pytorch/pull/65851/files#diff-0fd0c99904c62a1a7e46b08568d047c4e9ab153b529a8c6a2317ac6a3726bdffL80

Here, the negation is happening inside the loop, and we are the mercy of the optimizer to lift it out of the loop evaluation. I don't remember if I checked if it was getting optimized this way or not.